util/src/main/java/org/killbill/billing/util/listener/RetryableSubscriber.java

```diff
@@ -0,0 +1,101 @@
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.reflect.TypeToken;
+
+public class RetryableSubscriber extends RetryableHandler {
+
+    private static final Logger log = LoggerFactory.getLogger(RetryableSubscriber.class);
+
+    public RetryableSubscriber(final Clock clock,
+                               final RetryableService retryableService,
+                               final NotificationQueueHandler handlerDelegate,
+                               final InternalCallContextFactory internalCallContextFactory) {
+        super(clock, retryableService, handlerDelegate, internalCallContextFactory);
+    }
+
+    public void handleEvent(final BusEvent event) {
+        handleReadyNotification(new SubscriberNotificationEvent(event, event.getClass()),
+                                clock.getUTCNow(),
+                                event.getUserToken(),
+                                event.getSearchKey1(),
+                                event.getSearchKey2());
+    }
+
+    public interface SubscriberAction<T extends BusEvent> {
+
+        void run(T event);
+    }
+
+    public static final class SubscriberQueueHandler implements NotificationQueueHandler {
+
+        // Similar to com.google.common.eventbus.SubscriberRegistry
+        private static final LoadingCache<Class<?>, ImmutableSet<Class<?>>> FLATTEN_HIERARCHY_CACHE =
+                CacheBuilder.newBuilder()
+                            .build(
+                                    new CacheLoader<Class<?>, ImmutableSet<Class<?>>>() {
+                                        @Override
+                                        public ImmutableSet<Class<?>> load(final Class<?> concreteClass) {
+                                            return ImmutableSet.<Class<?>>copyOf(TypeToken.of(concreteClass).getTypes().rawTypes());
+                                        }
+                                    });
+
+        private final Map<Class<?>, SubscriberAction<? extends BusEvent>> actions = new HashMap<Class<?>, SubscriberAction<? extends BusEvent>>();
+
+        public SubscriberQueueHandler() {}
+
+        public <B extends BusEvent> void subscribe(final Class<B> busEventClass, final SubscriberAction<B> action) {
+            actions.put(busEventClass, action);
+        }
+
+        @Override
+        public void handleReadyNotification(final NotificationEvent eventJson, final DateTime eventDateTime, final UUID userToken, final Long searchKey1, final Long searchKey2) {
+            if (!(eventJson instanceof SubscriberNotificationEvent)) {
+                log.error("SubscriberQueueHandler received an unexpected event className='{}'", eventJson.getClass());
+            } else {
+                final BusEvent busEvent = ((SubscriberNotificationEvent) eventJson).getBusEvent();
+
+                final ImmutableSet<Class<?>> eventTypes = FLATTEN_HIERARCHY_CACHE.getUnchecked(busEvent.getClass());
```

Comment:
In our use case, we could even consider simplifying the code, because i don't think we have use cases where handlers are `@Subscribing`  to base class/super-interface events.